ES6: Convert modules/component

gustavpursche opened the pull request:
woop woop

SiAdcock commented on static/src/javascripts/projects/admin/modules/abtests/participation-item.js:
```diff
@@ -23,13 +23,17 @@ class ParticipationItem extends Component {
         const href = `${this.config.examplePath}=${this.config.variant}`;
-        this.elem.textContent = this.config.variant;
-        this.elem.href = origin + href;
+        const elem = ((this.elem: any): HTMLLinkElement);
```
To avoid this cast, shall we make the type of `this.elem: ?HTMLElement | ?HTMLLinkElement`

?

gustavpursche replied:
Yeah, we could. I'm afraid it would become a growing list, but I think you're right. 👍

SiAdcock commented on static/src/javascripts/projects/common/modules/component.js:
```diff
@@ -0,0 +1,366 @@
+
+export { Component };
+
+/* eslint-disable no-underscore-dangle */
```
Does this need to be repeated here?

SiAdcock commented on static/src/javascripts/projects/common/modules/discussion/activity-stream.js:
```diff
@@ -129,15 +134,20 @@ class ActivityStream extends Component {
     }
 
     updateHistory(resp: Object): void {
-        const page = this.options.page;
```
```js
const options = this.options || {};
```
Would save at least 4 existential checks in this function!

SiAdcock commented on static/src/javascripts/projects/common/modules/discussion/comment-box.js:
```diff
@@ -215,8 +224,11 @@ class CommentBox extends Component {
 
     postComment(): void {
+        // $FlowFixMe
+        const { value } = this.elem && this.elem.body;
```
I guess Flow's unhappy here because if `this.elem` doesn't exist we'd get a `TypeError: Cannot match against 'undefined' or 'null'`? Should we refactor to be on the safe side?

gustavpursche replied:
Not quite: https://github.com/guardian/frontend/pull/18037/commits/597738d2bbe9682c4371694758c79b0fb1e9db2f

SiAdcock commented on static/src/javascripts/projects/common/modules/component.js:
```diff
@@ -0,0 +1,366 @@
+        bean.fire(this.elem, eventName, args);
+    }
+
+    getElem(elemName: string): HTMLElement {
```
The return type of this method is being junked in a number of places by downcasting to `any` then back up to `HTMLXXXXElement`. Should we change the return type to account for the many possibilities, e.g.

```
getElem(elemName: string): HTMLElement | HTMLLinkElement | HTMLInputElement
```

And perform refinements where necessary

```js
const commentBody = ((this.getElem('body'): any): HTMLInputElement);

if (
  commentBody instanceof 'HTMLInputElement' &&
  commentBody.value.length === 0
) {
  // ...
}
```
I think this is more type-safe approach than casting to `any`. What do you think?

SiAdcock replied:
Just spoke asynchronously with @GHaberis about this, and we both came to the conclusion that `getElem`'s return type should be nullable, as it is possible that the selector won't match anything.

Do you think this would be a total drag, with lots of null checks everywhere? Saying that, we should bear in mind that having lots of null checks has dramatically reduced the number of errors on the site.

I'd be happy to discuss this in person tomorrow if you prefer.

GHaberis replied:
Just seconding Simon comment above ☝️ . I did wonder looking at this whether it's misleading that the return type was `HTMLElement` when it feasibly could be `null`. If we swapped out `qwery` for vanilla JS `flow` would warn about this. As Simon says making the return type nullable will mean we'll need to wrap references to elems derived from `getElem` in null checks, but this would be safer...but yeah it could be a pain to do this.

GHaberis commented on static/src/javascripts/projects/common/modules/discussion/comment-box.js:
```diff
@@ -186,13 +192,16 @@ class CommentBox extends Component {
     postCommentSuccess(comment: commentType, resp: Object): void {
-        if (this.options.newCommenter) {
-            this.refreshUsernameHtml();
+        this.refreshUsernameHtml();
```
Is it OK to call this if `this.options.newCommenter` is falsy?

GHaberis commented on static/src/javascripts/projects/common/modules/discussion/comment-box.js:
```diff
@@ -100,16 +100,19 @@ class CommentBox extends Component {
-                .getAttribute('data-discussion-key')
-                .replace('discussion', '')
-        );
+        const discussionId = this.options && this.options.discussionId;
```
Could we just return `discussionId` after this line if it exists? All the code beneath isn't required to run if `discussionId` is defined.

PRBuilds commented:
PRbuilds results:


**Screenshots**
[desktop.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/desktop.png) • [tablet.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/tablet.png) • [mobile.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/mobile.png) • [wide.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/wide.png)



&#x1F49A; **Exceptions**
[thrown-exceptions.js](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/thrown-exceptions.js)





&#x1F525; (failed)  **A11y validation**
[a11y-report.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/a11y-report.txt)



**Apache Benchmark Load Testing**
[loadtesting.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/loadtesting.txt)



&#x1F525; (invalid)  **Microdata Validation**
[microdata.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/microdata.txt)

    
--automated message

gustavpursche commented:
@GHaberis @SiAdcock I think I addressed all your comments. Would be great to have another 👀, so that I can start click-testing after on the website, to see if everything still works as expected :)

PRBuilds commented:
PRbuilds results:


**Screenshots**
[desktop.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/desktop.png) • [tablet.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/tablet.png) • [mobile.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/mobile.png) • [wide.png](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/screenshots/books%252F2014%252Fmay%252F21%252Fguardian-journalists-jonathan-freedland-ghaith-abdul-ahad-win-orwell-prize-journalism/wide.png)



&#x1F49A; **Exceptions**
[thrown-exceptions.js](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/thrown-exceptions.js)





&#x1F525; (failed)  **A11y validation**
[a11y-report.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/a11y-report.txt)



**Apache Benchmark Load Testing**
[loadtesting.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/loadtesting.txt)



&#x1F525; (invalid)  **Microdata Validation**
[microdata.txt](https://s3-eu-west-1.amazonaws.com/prbuilds/PR-18037/microdata.txt)

    
--automated message

gustavpursche commented:
🌴 Green build. 💌 specs.

gustavpursche commented:
@SiAdcock @GHaberis Tested the whole PR a bit by clicking around, but couldn't find any breakages. If you like, feel good about it and want to have an eye on it during the day, you could merge it tomorrow morning. Otherwise, let's do it on monday.
